scio-extra/src/main/scala/com/spotify/scio/extra/annoy/AnnoyUri.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2017 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.extra.annoy
+
+import java.io.File
+import java.net.URI
+import java.nio.file.{Files, Paths}
+
+import annoy4s._
+import com.spotify.scio.util.{RemoteFileUtil, ScioUtil}
+import com.sun.jna.Native
+import org.apache.beam.sdk.extensions.gcp.options.GcsOptions
+import org.apache.beam.sdk.options.PipelineOptions
+
+/**
+ * Represents the base URI for an Annoy tree, either on the local or a remote file system.
+ */
+trait AnnoyUri extends Serializable {
+  val path: String
+  private[annoy] def getReader(metric: ScioAnnoyMetric, dim: Int, nTrees:Int): AnnoyReader
+  private[annoy] def saveAndClose(annoyIndex: AnnoyWriter): Unit
+  private[annoy] def exists: Boolean
+  override def toString: String = path
+}
+
+private[annoy] object AnnoyUri {
+
+  def apply(path: String, opts: PipelineOptions): AnnoyUri =
+    if (ScioUtil.isLocalUri(new URI(path))) {
+      new LocalAnnoyUri(path)
+    } else {
+      new RemoteAnnoyUri(path, opts.as(classOf[GcsOptions]))
+    }
+
+}
+
+private class LocalAnnoyUri(val path: String) extends AnnoyUri {
+
+  override private[annoy] def getReader(metric: ScioAnnoyMetric, dim: Int, nTrees: Int): AnnoyReader
+  = new AnnoyReader(metric, dim, nTrees, path)
+  override private[annoy] def saveAndClose(w: AnnoyWriter): Unit = {
+    w.build
+    w.save(path.toString)
+    w.free()
```

Comment:
Is this sth that should be in `finally`?

Reply:
Like this `try { w.build(); w.save(path.toString) } finally { w.free() }`?